test/js/bun/shell/test_builder.ts

```diff
@@ -260,63 +304,104 @@ export function createTestBuilder(path: string) {
-    //     if (tb.expected_stderr !== undefined) expect(stderr.toString()).toEqual(tb.expected_stderr);
-    //     if (tb.expected_exit_code !== undefined) expect(exitCode).toEqual(tb.expected_exit_code);
-
-    //     for (const [filename, expected] of Object.entries(tb.file_equals)) {
-    //       const actual = await Bun.file(filename).text();
-    //       expect(actual).toEqual(expected);
-    //     }
-    //     return output;
-    //   }
-
-    //   if (this._testName !== undefined) {
-    //     test(this._testName, async () => {
-    //       await doTest(this);
-    //     });
-    //   }
-    //   await doTest(this);
-    // }
-  }
-  function generateRandomString(length: number): string {
-    const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
-    let result = "";
-    const charactersLength = characters.length;
+    generateRandomString(length: number): string {
+      const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
+      let result = "";
+      const charactersLength = characters.length;
 
-    for (let i = 0; i < length; i++) {
-      result += characters.charAt(Math.floor(Math.random() * charactersLength));
+      for (let i = 0; i < length; i++) {
+        result += characters.charAt(Math.floor(Math.random() * charactersLength));
+      }
+
+      return result;
     }
 
-    return result;
+    joinTemplate(): string {
+      let buf = [];
+      for (let i = 0; i < this._scriptStr.length; i++) {
+        buf.push(this._scriptStr[i]);
+        if (this._expresssions[i] !== undefined) {
+          const expr = this._expresssions[i];
+          this.processShellExpr(buf, expr);
+        }
+      }
+
+      return buf.join("");
+    }
+
+    processShellExpr(buf: string[], expr: ShellExpression) {
+      if (typeof expr === "string") {
+        buf.push(Bun.$.escape(expr));
+      } else if (typeof expr === "number") {
+        buf.push(expr.toString());
+      } else if (typeof expr?.raw === "string") {
+        buf.push(Bun.$.escape(expr.raw));
+      } else if (Array.isArray(expr)) {
+        expr.forEach(e => this.processShellExpr(buf, e));
+      } else {
+        if (this._skipExecOnUnknownType) {
+          console.warn(`Unexpected expression type: ${expr}\nSkipping.`);
+          return;
+        }
+        throw new Error(`Unexpected expression type ${expr}`);
+      }
+    }
   }
 
   return TestBuilder;
 }
+
+function generateRandomString(length: number): string {
+  const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
+  let result = "";
+  const charactersLength = characters.length;
+
+  for (let i = 0; i < length; i++) {
+    result += characters.charAt(Math.floor(Math.random() * charactersLength));
+  }
+
+  return result;
+}
+
+export function bunExe() {
+  if (process.platform === "win32") return process.execPath.replaceAll("\\", "/");
+  return process.execPath;
+}
```

Comment:
don't we already have a `bunExe` function in `harness.ts`?

Reply:
the shell leak tests do some hacky things like taking the `test_builder.ts` file and copying it, its easier to just rewrite the function here